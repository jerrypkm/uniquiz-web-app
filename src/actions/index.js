export const addToState = (store, data) => {
  const state = {
    ...store.state,
    ...data,
  };
  store.setState(state);
};
export const addToCollections = (store, data) => {
  const collections = {
    ...store.state.collections,
    ...data,
  };
  store.setState({ collections });
};

export const addToAuthUser = (store, data) => {
  let authUser = {
    ...store.state.authUser,
    ...data,
  };

  if (!data) {
    authUser = null;
  }
  store.setState({ authUser });
};

export const addToUserRole = (store, data) => {
  let userRole = {
    ...store.state.userRole,
    ...data,
  };

  if (!data) {
    userRole = null;
  }
  store.setState({ userRole });
};

export const addToFirebase = (store, data) => {
  const firebaseCache = {
    ...store.state.firebaseCache,
    ...data,
  };
  store.setState({ firebaseCache });
};

export const addToCourses = (store, data) => {
  const courses = {
    ...store.state.courses,
    ...data,
  };
  store.setState({ courses });
};

export const addToThemeOptions = (store, data) => {
  const themeOptions = {
    ...store.state.themeOptions,
    ...data,
  };
  store.setState({ themeOptions });
};
