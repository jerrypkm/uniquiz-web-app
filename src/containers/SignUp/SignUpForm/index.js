import React, { useState } from "react";

import * as ROLES from "constants/roles";

import { compose } from "recompose";
import EmailSignUpForm from "components/shared/EmailSignUpForm";
import { withFirebase } from "components/shared/Firebase";
import { withRouter } from "react-router-dom";
import withStyles from "@material-ui/core/styles/withStyles";

const styles = (theme) => ({
  button: {
    margin: theme.spacing(1),
  },
  input: {
    margin: theme.spacing(1),
  },
});

const INITIAL_STATE = {
  username: "",
  email: "",
  passwordOne: "",
  passwordTwo: "",
  isAdmin: false,
  error: null,
};

const ERROR_CODE_ACCOUNT_EXISTS = "auth/email-already-in-use";

const ERROR_MSG_ACCOUNT_EXISTS = `
  An account with this E-Mail address already exists.
  Try to login with this account instead. If you think the
  account is already used from one of the social logins, try
  to sign in with one of them. Afterward, associate your accounts
  on your personal account page.
`;

const SignUpFormBase = (props) => {
  const [state, setState] = useState({ ...INITIAL_STATE });
  const { username, email, passwordOne, passwordTwo, error, isAdmin } = state;

  const handleSubmit = (event) => {
    const roles = {};

    if (isAdmin) {
      roles[ROLES.ADMIN] = ROLES.ADMIN;
    }

    props.firebase
      .createUserWithEmailAndPassword(email, passwordOne)
      .then((authUser) => {
        // Create a user in your Firebase realtime database
        return props.firebase.user(authUser.user.uid).set(
          {
            username,
            email,
            roles,
          },
          { merge: true }
        );
      })
      .then(() => {
        return props.firebase.sendEmailVerification();
      })
      .then(() => {
        setState({ ...INITIAL_STATE });
        props.history.goBack();
      })
      .catch((error) => {
        if (error.code === ERROR_CODE_ACCOUNT_EXISTS) {
          error.message = ERROR_MSG_ACCOUNT_EXISTS;
        }
        setState({ ...state, error });
      });

    event.preventDefault();
  };

  const onChange = (event) => {
    setState({ ...state, [event.target.name]: event.target.value });
  };

  return (
    <EmailSignUpForm
      onSubmit={handleSubmit}
      isInvalid={
        passwordOne !== passwordTwo ||
        passwordOne === "" ||
        email === "" ||
        username === ""
      }
      onChange={onChange}
      error={error}
    />
  );
};

const SignUpForm = compose(
  withStyles(styles),
  withRouter,
  withFirebase
)(SignUpFormBase);

export default SignUpForm;
