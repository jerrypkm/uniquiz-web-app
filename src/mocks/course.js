export const course = {
  id: "123",
  question: [],
  title: "Test Course",
};

export const collections = {
  courses: [
    {
      id: 0,
      isDisabled: false,
      question: [],
      title: "Html",
      type: "html",
      subtitle: "Learn Html",
      doc: "html",
    },
    {
      id: 1,
      isDisabled: false,
      question: [],
      title: "JavaScript",
      type: "js",
      subtitle: "Learn JavaScript",
      doc: "js",
    },
    {
      id: 2,
      isDisabled: true,
      question: [],
      title: "React",
      type: "reactJsx",
      subtitle: "Learn React",
      doc: "react",
    },
  ],
  "courses/html/questions": [
    {
      id: "1",
      title: "Html",
      subtitle: "Learn Basic Html Coding",
      questions: [
        {
          id: "1",
          label: "hello",
          title: "hello",
          question: "Hello World",
          answer: "<p>Hello React</p>",
        },
      ],
    },
  ],
  "courses/js/questions": [
    {
      id: "1",
      title: "JavaScript",
      subtitle: "Learn Basic JavaScript in JSX",
      questions: [{ id: "1" }],
    },
  ],
  "courses/react/questions": [
    {
      id: "1",
      title: "React",
      subtitle: "Learn Basic to Advanced ",
      questions: [{ id: "1" }],
    },
  ],
};
