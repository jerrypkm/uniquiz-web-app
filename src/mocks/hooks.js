import React from "react";

export const useStateMock = jest
  .spyOn(React, "useState")
  .mockImplementation(() => jest.fn());
