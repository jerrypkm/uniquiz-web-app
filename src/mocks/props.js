export const props = {
  match: { params: { questionId: "1", collection: "test-collection" } },
  history: { location: { hash: "#" }, push: () => {} },
};
