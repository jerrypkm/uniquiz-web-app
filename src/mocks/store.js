// Mock functionality of global and store hooks

export const storeMock = jest.mock("store", () =>
  jest.fn().mockReturnValue([{ state: { authUser: null } }])
);
export const globalHookMock = jest.mock("use-global-hook", () => jest.fn([]));
