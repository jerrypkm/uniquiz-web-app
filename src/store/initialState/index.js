import { getCache } from "helpers/localCache";

const isDarkMode = getCache({ key: "isDarkMode" }) || false;

export const initialState = {
  authUser: null,
  collections: {},
  courses: [],
  themeOptions: {
    isDarkMode,
  },
  firebase: null,
};
