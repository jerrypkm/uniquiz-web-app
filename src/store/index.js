import React from "react";

import * as actions from "actions";
import { initialState } from "./initialState";
import useGlobalHook from "use-global-hook";

const useGlobal = useGlobalHook(React, initialState, actions);

export default useGlobal;
