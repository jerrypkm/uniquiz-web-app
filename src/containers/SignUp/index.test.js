import React from "react";
import SignUpForm from "./index";

import { initialState } from "store/initialState";

// Dependencies under test
import ShallowRenderer from "react-test-renderer/shallow";
import useGlobal from "store";
import useGlobalHook from "use-global-hook";

import { renderHook } from "@testing-library/react-hooks";

import actions from "actions";
import { props } from "mocks/props";

// Shallow render to test the component without it's children
const renderer = new ShallowRenderer();

// Mock functionality of global and store hooks
jest.mock("store", () =>
  jest.fn().mockReturnValue([{ state: { authUser: null } }])
);
jest.mock("use-global-hook", () => jest.fn([]));

beforeEach(() => {
  useGlobal.mockClear();
  useGlobalHook.mockClear();
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe("SignUpForm Component", () => {
  test("Should import and invoke useGlobal", () => {
    // Call useGlobal hook
    const { result, waitForNextUpdate } = renderHook(() =>
      useGlobal(React, initialState, actions)
    );
    // Wait for renders
    waitForNextUpdate();
    expect(result.current).toBeDefined();
    expect(useGlobal).toHaveBeenCalledWith(React, initialState, actions);
  });

  test("should match the snapshot", () => {
    renderer.render(<SignUpForm {...props} />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
