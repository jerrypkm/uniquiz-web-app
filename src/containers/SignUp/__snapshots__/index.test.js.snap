// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm Component should match the snapshot 1`] = `
<ForwardRef(WithStyles)
  maxWidth="lg"
>
  <ForwardRef(WithStyles) />
  <ForwardRef(WithStyles)
    isLoading={false}
    title="Sign up to get started ðŸ˜Ž"
  >
    <ForwardRef(WithStyles)
      container={true}
      spacing={3}
    >
      <React.Fragment>
        <ForwardRef(WithStyles)
          item={true}
          md={6}
          sm={6}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <ForwardRef(WithStyles) />
        </ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          item={true}
          md={6}
          sm={6}
        >
          <ForwardRef(WithStyles) />
        </ForwardRef(WithStyles)>
      </React.Fragment>
    </ForwardRef(WithStyles)>
  </ForwardRef(WithStyles)>
</ForwardRef(WithStyles)>
`;
